### PR TITLE
Rewriting the CI recipes to make use of mamba + cache

### PR DESCRIPTION
**Pull request by ThomasLecocq**
# What does this PR do?
- It implements a new way to create the environments for CI testing on github-actions.
- The recipes have been modified to install mamba & to save/reuse daily snapshots of the environments.
- There is also a new job that checks for existing "running" or "queued" jobs in _the same PR_ (!!) and cancels them, to save CI energy for the latest push, and discard outdated content.

Using cached environments saves at least 2 minutes on Ubuntu, could save up to 10 min on macOS & Windows for the environment solving step. This comes at a price: every day, the first CI run will need to cache the environment (tar-gz it and save it remotely, which is almost instantaneous on *nux, takes 2min on windows).

### CODE REVIEW

**Comment by ThomasLecocq** [comment]
So, that was the first run: 
image
will now trigger a new build with a changelog entry

**Comment by ThomasLecocq** [comment]
made a mistake previously, now the cache is properly created:
image

**Comment by ThomasLecocq** [comment]
After caching:
image
It's not amaaaazing... 

**but:**

Ubuntu tests start with a lead time of about 85 seconds, compared to 197 seconds before.
If I re-enable the "fail tests on first matrix element failure", it'd mean the ubuntu results should already let users know something is wrong with their PR within less than 4 to 6 minutes!

**Comment by ThomasLecocq** [comment]
I added a job canceller in case we push a new commit to the same PR, to avoid loosing test-minutes on outdated content.

**Comment by ThomasLecocq** [comment]
question to @obspy/maintainers should the other matrix elements be cancelled when 1 fails (e.g.: ubuntu 3.8 fails, should the macOS and windows (usually slower) be cancelled, as it's "useless to run them", or should they be allowed to run, to see if the error is only on ubuntu 3.8 ?

**Comment by d-chambers** [comment]
Probably let them run, why not?

**Comment by claudiodsf** [comment]
I'm for letting other jobs run: it's important to know if failure is platform-related.

**Review by claudiodsf — APPROVED** [review]
Fine for me if all the tests are green.

I would suggest to squash all the commits before merging.

**Comment by ThomasLecocq** [comment]
rebased & ready to squash-merge when 🟢

**Comment by ThomasLecocq** [comment]
merged 🟢 , let's see what master test complain about :-)

**Comment by claudiodsf** [comment]
But I would have preferred to squash commits before 🙄

**Comment by ThomasLecocq** [comment]
why ? it's squashed on master

**Comment by ThomasLecocq** [comment]
🟢 all tests work on master, me happy :-)

**Comment by claudiodsf** [comment]
> why ? it's squashed on master (I always do "Squash & Merge" when merging)

Ah ok, didn't know that 🙄

**Pull request merged**
